Implement mock crypto and symmetric encryption functions [MAID-2702]

nbaksalyar opened the pull request:
Implements fake `rust_sodium` without real crypto for testing purposes

nbaksalyar committed:
chore/license: fix incorrect license header

nbaksalyar committed:
feat/mock-crypto: add optional mock-crypto feature

nbaksalyar committed:
chore/ci: add mock-crypto testing for CI

nbaksalyar committed:
feat/mock-crypto: disable authenticated encryption test

It depends on having real crypto for precomputed keys.

ustulation commented on src/lib.rs:
```diff
@@ -35,8 +36,13 @@
 #![allow(missing_docs)]
```
why not add docs ?

ustulation commented on src/mock_crypto.rs:
```diff
@@ -0,0 +1,242 @@
+
+            /// Generate mock shared key
+            pub fn precompute(pk: &PublicKey, _sk: &SecretKey) -> PrecomputedKey {
+                PrecomputedKey(pk.0)
```
XOR of 2 unique numbers will always be unique - so why not use that and utilise both (their) pk and (our) sk above to generated the shared secret ?

ustulation commented on src/lib.rs:
```diff
@@ -259,6 +272,9 @@ mod tests {
 
+    // We don't run this test with mock-crypto because it depends on enabling real crypto features
+    // to use precomputed keys.
+    #[cfg(not(feature = "use-mock-crypto"))]
```
why ? we can do it via [review comment to this](https://github.com/maidsafe/safe_crypto/pull/5/files#diff-09b0eb261ff3c84c9864b11431935b45R150) - or will that be a problem somehow ?

ustulation commented:
[This one](https://github.com/maidsafe/safe_crypto/pull/5/files#diff-b4aea3e418ccdb71239b96952d9cddb6R159) is misleading - better to call it `shared_secret()` ? Else it sounds like it's a shareable key.

ustulation commented:
No facility to use Symmetric cipher. So either that must be provided OR if it's the same tied to SharedSecretKey (investigate the lib being used and see if SharedSecretKey can be used for Symmetric Ciphers (i.e. output of something like `gen-symmetric-cipher-key` can be force fed into SharedSecretKey without changing the underlying mech.

ustulation commented on src/mock_crypto.rs:
```diff
@@ -0,0 +1,242 @@
+                let n = nonce.0.len();
+                let s = sk.0.len();
+
+                if c[0..n] != nonce.0 {
```
why is this done only for nonce and not for secret-key ?

nbaksalyar replied:
Should be fixed with a new commit.

ustulation commented on src/mock_crypto.rs:
```diff
@@ -0,0 +1,242 @@
+            }
+
+            /// Perform mock anonymous decryption.
+            pub fn open(c: &[u8], pk: &PublicKey, _sk: &SecretKey) -> Result<Vec<u8>, ()> {
```
since in gen_keypair for mock we derive PK==SK we can do the additional assert here to make sure the correct keys are passed - so it should result in error if pk != sk that way you'll use both the passed keys.

ustulation commented on src/mock_crypto.rs:
```diff
@@ -0,0 +1,242 @@
+            use rand::Rng;
+
+            /// Number of bytes in a `PublicKey`.
+            pub const PUBLICKEYBYTES: usize = 32;
```
Do we need these to be this big ? Can 8 be not sufficient ? So everything in mock is `[u8; 8]` for e.g.

ustulation commented on src/mock_crypto.rs:
```diff
@@ -0,0 +1,242 @@
+            use std::ops::{Index, RangeFull};
+
+            /// Number of bytes in a `PublicKey`.
+            pub const PUBLICKEYBYTES: usize = 32;
```
Do we need these to be this big ? Can 16 be not sufficient ? So everything in mock is [u8; 16 for e.g. since that is the lowest keccak can output (sha3_128) i think. Or may use a non-crypto but fast hasher like [this one](https://doc.servo.org/fnv/struct.FnvHasher.html) and get a `u64` vai [finish](https://doc.servo.org/core/hash/trait.Hasher.html#tymethod.finish) and then have that in [u8; 8] which will be similar to the `box_` counterparts below and probably faster than tiny-keccak.

These are all suggestions ofc, do only if you don't see a problem with it and can work with seeded-rng to reprocduce failures.

nbaksalyar replied:
It seems that `FnvHasher` is not relevant anymore. According to the [benchmark it refers](https://github.com/Gankro/hash-rs), `DefaultHasher` from the standard lib is much faster. E.g., compare:
```
test fnv::bytes_000002048  ... bench:       1,949 ns/iter (+/- 29) = 1050 MB/s
test sip::bytes_000002048  ... bench:         429 ns/iter (+/- 3) = 4773 MB/s
```
or
```
test fnv::mapcountdense_000002048    ... bench:   2,020,220 ns/iter (+/- 51,216) = 1013 MB/s
test fnv::mapcountsparse_000002048   ... bench:   2,005,899 ns/iter (+/- 43,826) = 1020 MB/s

test sip::mapcountdense_000002048    ... bench:     496,136 ns/iter (+/- 5,623) = 4127 MB/s
test sip::mapcountsparse_000002048   ... bench:     493,424 ns/iter (+/- 4,971) = 4150 MB/s
```
(`sip` is the default hasher, `fnv` is `FnvHasher`)

So I'd probably try replacing `tiny_keccak` with the default hasher.

nbaksalyar committed:
fix/mock-crypto: use XOR to compute shared secret key

nbaksalyar committed:
feat/docs: add documentation for public methods and structs

nbaksalyar committed:
refactor/naming: change `shared_key` to `shared_secret`

The latter name better represents the purpose of the method.

nbaksalyar committed:
refactor/mock-crypto: use non-crypto hasher for mock

Replace tiny_keccak with DefaultHasher for signature purposes.
This removes unnecessary dependency and improves the performance.

nbaksalyar committed:
feat/symmetric: add symmetric encryption functions

Including the mock version

ustulation commented on src/mock_crypto.rs:
```diff
@@ -88,19 +90,31 @@ pub mod rust_sodium {
-                use tiny_keccak::sha3_256;
-                sha3_256(data)
+            fn hash128(data: &[u8]) -> [u8; 8] {
+                let mut hasher = DefaultHasher::new();
```
One of the points was it needs determinism - is this deterministic w.r.t given seeds else reproduction of failure will be impossible ?

Fraser999 commented on src/lib.rs:
```diff
@@ -1,10 +1,11 @@
-// KIND, either express or implied. Please review the Licences for the specific language governing
-// permissions and limitations relating to use of the SAFE Network Software.
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// http://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
```
https?

Fraser999 commented on src/lib.rs:
```diff
@@ -328,7 +505,37 @@ mod tests {
+        assert_eq!(plaintext, data);
+    }
+
+    fn generate_random_string(length: usize) -> Vec<u8> {
```
Would this be clearer as `generate_random_bytes`?

Fraser999 commented on src/mock_crypto.rs:
```diff
@@ -0,0 +1,323 @@
+// Copyright 2018 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// http://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
```
https?

Fraser999 commented on Cargo.toml:
```diff
@@ -15,6 +15,10 @@ rust_sodium = "~0.9.0"
 serde = "~1.0.66"
 serde_derive = "~1.0.66"
 unwrap = "~1.1.0"
+rand = { version = "~0.5.2", optional = true }
```
This version of `rand` will mean we can't use the current [`SeededRng`](https://docs.rs/maidsafe_utilities/0.15.0/maidsafe_utilities/struct.SeededRng.html) provided by maidsafe_utilities.  Not sure if that's a problem though?

Fraser999 commented on src/mock_crypto.rs:
```diff
@@ -0,0 +1,323 @@
+            use std::ops::{Index, RangeFull};
+
+            /// Number of bytes in a `PublicKey`.
+            pub const PUBLICKEYBYTES: usize = 8;
```
Having these `const`s different to the real ones could be problematic too I think.  We wouldn't for example be able to use the same code to implement the XOR of a SHA256 with a mock and non-mock `sign::PublicKey`.

Having said that, I'm not sure if or how we're planning to implement something like `safe_crypto::PublicId::name(&self) -> [u8; 32]`, so I guess that difference could be handled inside there.  This could be a non-issue too.

Fraser999 commented on src/mock_crypto.rs:
```diff
@@ -0,0 +1,323 @@
+                *signature == Signature(hash128(&temp))
+            }
+
+            fn hash128(data: &[u8]) -> [u8; 8] {
```
This isn't returning 128 bits, only 64, so I'd suggest renaming the function.

Fraser999 commented on src/lib.rs:
```diff
@@ -45,17 +47,28 @@ extern crate quick_error;
-use rust_sodium::crypto::{box_, sealedbox, sign};
+use rust_sodium::crypto::{box_, sealedbox, secretbox, sign};
+#[cfg(feature = "use-mock-crypto")]
+pub use rust_sodium::{init as mock_crypto_init, init_with_rng as mock_crypto_init_with_rng};
```
Shouldn't these be abstracted away too?  I'd have thought we'd provide our own `init()` and `init_with_rng()` functions in the mock and non-mock API, and internally these can call the rust_sodium equivalents.

I don't think we'd be able to just wrap the `init_with_rng()` however as long as we have the rand version mismatch.  We'd only be able to expose that for the mock configuration unless we update rust_sodium's version of rand too.

Fraser999 commented on src/mock_crypto.rs:
```diff
@@ -0,0 +1,323 @@
+    }
+
+    /// Initialise mock `rust_sodium`.
+    pub fn init() -> bool {
```
The real [`init()`](https://docs.rs/rust_sodium/0.9.0/rust_sodium/fn.init.html) and [`init_with_rng()`](https://docs.rs/rust_sodium/0.9.0/rust_sodium/fn.init_with_rng.html) functions have different return types to those here.

I guess that's not a real problem if we hide them behind our own API functions which call these internally (although those functions would internally be doing something slightly different for mock vs non-mock), but if we expose these through our API by just directly exporting them, I think we should make sure their signatures match the real ones' signatures.

Fraser999 commented on src/mock_crypto.rs:
```diff
@@ -0,0 +1,323 @@
+//! they are NOT secure. They are supposed to be used for testing only.
+
+/// Mock version of a subset of the `rust_sodium` crate.
+pub mod rust_sodium {
```
Making every `pub` in this file `pub(crate)` would make the scope of these clearer and help avoid any of them being accidentally exported in the future?  (We'd need an exception for the two init functions if they are to be exported).

Fraser999 commented on src/lib.rs:
```diff
@@ -269,37 +446,37 @@ mod tests {
 
         // Trying with wrong data
-        let error_res: Result<_, _> = shared_key2.decrypt_bytes(&plaintext);
+        let error_res: Result<_, _> = shared_sk2.decrypt_bytes(&plaintext);
```
This (and a few other instances) could be cut down a bit to:
```rust
let _ = shared_sk2.decrypt_bytes(&plaintext).expect_err(
    "Unexpected result: we're using wrong data, it should have returned an error",
);
```

nbaksalyar replied:
Ah great, didn't know this was added to `Result` :+1: Thanks!

Fraser999 commented on src/mock_crypto.rs:
```diff
@@ -0,0 +1,323 @@
+
+                let hash = hasher.finish();
+                [
+                    ((hash >> 56) & 0xFF) as u8,
```
They're not doing any harm, but I'm pretty sure we don't need any of the `& OxFF` masks here.

nbaksalyar committed:
refactor/all: address review points

- Fix license links in the header
- Rename functions to better reflect their purpose
- Make mock-crypto functions package private
- Simplify tests code
- Simplify mock-crypto code
- Add `Default` impl for SymmetricKey

nbaksalyar committed:
fix/deps: downgrade to rand 0.4

This is required because maidsafe_utilities and other crates
still depend on rand 0.4. Eventually they'll be upgraded to
rand 0.5 but for now we need to maintain the compatibility.

nbaksalyar committed:
feat/init: provide initialisation functions

nbaksalyar committed:
feat/id: provide the function to get the public ID name

ustulation commented on src/lib.rs:
```diff
@@ -120,6 +120,29 @@ struct CipherText {
+            .cloned()
+            .cycle()
+            .take(32)
+            .collect::<Vec<u8>>();
```
Why allocate stuff on heap only to throw it away ? Why not something like:
```
take(32).enumerate().for_each(|(i, v)| key_array[i] = v);
```

ustulation commented on src/lib.rs:
```diff
@@ -387,29 +393,21 @@ mod tests {
-        }
+        let _ = sk
+            .decrypt_anonymous_bytes(&data)
+            .expect_err("Unexpected result: we're using wrong data, it should have returned error");
```
@nbaksalyar Sorry man, i guess these came from review that @Fraser999 added and i overlooked this (incorrect) review point of his. `expect_err()` belongs to `unwrap()/expect()` etc family of functions which we deliberately avoid because it does not give us the line numbers etc and no backtrace in release. Hence the usage of macros like `unwrap!()/panic!()` - so what you did earlier was correct and i spoke to fraser and he agrees - also in line with our style guide. I would let this pass for this iteration but there's another review i added so makes sense to just address this anyway.

Fraser999 replied:
Sorry - my bad! :(

nbaksalyar committed:
refactor/all: address review points

- Get rid of needless heap alloc
- Revert the error reporting code (we should not use expect_err because it won't
  show a line number for a place where error occurred)
- Mark all objects in the mock_crypto module as pub(crate)
- Do not use alias to import rust_sodium::init and rust_sodium::init_with_rng
